statsmodels/tsa/holtwinters.py

```diff
@@ -0,0 +1,472 @@
+        try:
+            if alpha*beta==0.0: return max_seen
+            if beta > alpha or gamma > 1-alpha: return max_seen
+            for i in range(1, N):
+                l[i] = (y_alpha[i-1] / s[i-1]) + (alphac * (l[i-1] * b[i-1]**phi))
+                b[i] = (beta * (l[i] / l[i-1])) + (betac * b[i-1]**phi)
+                s[i+m-1] = (y_gamma[i-1] / (l[i-1] * b[i-1]**phi)) + (gammac * s[i-1])
+            return sqeuclidean((l*b**phi)*s[:-(m-1)], y)
+        except RuntimeWarning:
+            return max_seen
+        finally:
+            warnings.filterwarnings("once")
+    def holt_win_add_add_dam(x, xi):
+        alpha,beta,gamma,phi,alphac,betac,gammac = holt_win_init(x, xi)        
+        try:
+            if alpha*beta==0.0: return max_seen
+            if beta > alpha or gamma > 1-alpha: return max_seen
+            for i in range(1, N):
+                l[i] = (y_alpha[i-1]) - (alpha * s[i-1]) + (alphac * (l[i-1] + phi*b[i-1]))
+                b[i] = (beta * (l[i] - l[i-1])) + (betac * phi*b[i-1])            
+                s[i+m-1] = y_gamma[i-1] - (gamma * (l[i-1] + phi*b[i-1])) + (gammac * s[i-1])
+            return sqeuclidean((l+phi*b)+s[:-(m-1)], y)
+        except RuntimeWarning:
+            return max_seen
+        finally:
+            warnings.filterwarnings("once")
+    def holt_win_mul_add_dam(x, xi):
+        alpha,beta,gamma,phi,alphac,betac,gammac = holt_win_init(x, xi)        
+        try:
+            if alpha*beta==0.0: return max_seen
+            if beta > alpha or gamma > 1-alpha: return max_seen
+            for i in range(1, N):
+                l[i] = (y_alpha[i-1]) - (alpha * s[i-1]) + (alphac * (l[i-1] * b[i-1]**phi))
+                b[i] = (beta * (l[i] / l[i-1])) + (betac * b[i-1]**phi)
+                s[i+m-1] = y_gamma[i-1] - (gamma * (l[i-1] * b[i-1]**phi)) + (gammac * s[i-1])
+            return sqeuclidean((l*phi*b)+s[:-(m-1)], y)        
+        except RuntimeWarning:
+            return max_seen
+        finally:
+            warnings.filterwarnings("once")
+    if seasoning:
+        l0 = y[np.arange(N)%m==0].mean()
+        b0 = ((y[m:m+m]-y[:m])/m).mean() if trending else None
+        s0 = list(y[:m]/l0) if seasonal=='mul' else list(y[:m]-l0)
+    elif trending:
+        l0 = y[0]
+        b0 = y[1]/y[0] if trend=='mul' else y[1]-y[0]
+        s0 = []
+    else:
+        l0 = y[0]
+        b0 = None
+        s0 = []
+    if optimised:        
+        init_alpha = alpha if alpha is not None else 0.5/max(m,1)
+        init_beta = beta if beta is not None else 0.1*init_alpha
+        init_gamma = None
+        init_phi = phi if phi is not None else 0.99
+        if seasoning:            
+            init_gamma = gamma if gamma is not None else 0.05*(1-init_alpha) 
+            xi = np.array([alpha is None, beta is None, gamma is None, True, trending, phi is None and damped]+[True]*m)
+            if seasonal=='mul':
```

Comment:
More pythonic would be to stick these in a dict[(seasonal,trend)].  This is like C 👍